scripts/ci/ci_before_install.sh

```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+#
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+set -xeuo pipefail
+
+MY_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+export KUBERNETES_VERSION=${KUBERNETES_VERSION:=}
+# Required for K8s v1.10.x. See
+# https://github.com/kubernetes/kubernetes/issues/61058#issuecomment-372764783
+if [[ ! -z "${KUBERNETES_VERSION}" ]]; then
+    sudo mount --make-shared /
+    sudo service docker restart
+fi
+pip install --upgrade pip
```

Comment:
Yes. I think we can even pin version here.

Reply:
To avoid upgrading pip by upgraded base image by Travis (this script runs outside of the docker)

Reply:
For the 3.5/3.6 - I think that might be an option to mix it "Maybe we can mix a bit, so do Postgres with 3.5 and Mysql with 3.6". There is also a discussion in devlist I started with runing also 3.7 (and possibly 3.8b1) What do you think @Fokko  @ashb ?

Reply:
Adding in a _single_ build (i.e. not all three DB backends) for extra python versions is probably worth it - I'm not sure it's worth doing 3.6 and 3.7. I guess it depends what versions distros ship.

We should probably test against the versions shipped by Debian, Ubuntu, Centos/Amazon Linux?

Reply:
My suggestion would be to remove the whole `pip install --upgrade pip` since this is regulated in Docker image itself as well: https://github.com/docker-library/python/blob/master/3.6/stretch/Dockerfile#L68

Reply:
@fokko - The ci_before_install.sh is run in the host environment before docker compose is run so this pip install is run in the host. So it used to run all the pylint/doc check and other rather than tests. They are not run in docker to speed it up - but using local python on the host. This was a workaround for the long time to start the original airflow-ci docker image - it was always installing all the dependencies first before doing anything and it took a long time to even do pylint/doccheck etc.. 

Thus said - with the new Dockerfile approach, runing all the pylint/docs etc. should be as fast as running them with local python installation and then we might indeed remove any python installation on the host. Should be super-easy and I might try to do it now . WDYT @fokko ?

Reply:
I believe it is a good idea to run the pylint/doc stuff within the container as well since this will make local development easier. If people want to lint locally, it is easier for people to fire up the docker image, instead of having to fix their system docker to the same version as in the CI.

That part I understand, but I don't think upgrading pip is something that we should do, why just not leave as is and use the version that comes with the docker image. In the end, it will take time, and what is the benefit of updating pip?

Reply:
Yep. I will remove upgrading :). 

I  had a discussion today with the new member of our team @nuclearpinguin who worked on pylint changes and he struggled with having different version of pylint locally and seeing different version. I think it will be very useful to have pylint/docs run in docker and easily runnable for local development !

Reply:
👍 

Yes, I've seen it before as well. Having the ability to easily run pylint/docs kind of tooling inside of a container makes life much easier.

Reply:
Hey @Fokko @ashb - I fixed all the problems except the kubernetes tests (hopefully @gerardo might help with that). Everything else is green :) and nice:
https://travis-ci.org/apache/airflow/builds/546571775 . 

Please take another look.

In this latest version there are a few things improved:

- I found a way how to speed up the builds even if setup.py changes. Please take a look at Dockerfile modifications but basically I am pre-installing the dependencies from setup.py downloaded from Airflow master of github repo. This way the dependencies are already there when setup.py is added so changing setup.py will not invalidate the cache and only new/changed dependencies will be installed - that is few minutes saved when dependencies change. Still we will have the clean-cache build running daily with all caching disabled - this way we will detect any "fresh/full" installation problems, so there is no risk.

- All the pylint/docs/mypy/flake are now run inside Docker :). And they are fully reproducible locally as well without having to install anything (running ci_pylint.sh/ci_docs.sh etc. locally will download/rebuild the right docker image and run the same checks that are run on CI. No more diifferent results locally and in CI! @BasPH  -that might make our pylint experience much better. I also updated CONTRIBUTING.md to explain how to run those.

- Also @BasPH - I've added Pylint section to the CONTRIBUTING.md explaining the process we follow for pylint checks. I hope it might be useful.

Please take a look and hopefully Kubernetes will be solved within days and we can merge it finally.

Reply:
Any chance for the kubernetes PR @gerardo  :)? Or should I try to fix the kubernetes part without waiting ?

Reply:
Still ongoing @potiuk, I'm running into some issues with coreos inside docker-compose. I'm still working on it but I don't want to stop you from advancing. So if you want to keep moving forward, it's fine for me.

Reply:
My changes live in https://github.com/gerardo/airflow/commits/kind-ci if you want to have a look.

Reply:
Cool. I will take a look today/tomorrow and hope I can finally merge it.

Reply:
@gerardo - I think we can take it on after this one is merged :)